bin/build-helmet.js

```diff
@@ -1,42 +1,68 @@
 #!/usr/bin/env node
+import { promises as fs } from "fs";
 import * as path from "path";
 import { fileURLToPath } from "url";
 import rollupTypescript from "@rollup/plugin-typescript";
-import { writeRollup, withCommonJsFile } from "./helpers.js";
+import {
+  writeRollup,
+  withCommonJsFile,
+  withEsmFile,
+  renameFile,
+  finalizeCommonJs,
+  moveDir
+} from "./helpers.js";
 
 const thisPath = fileURLToPath(import.meta.url);
 const rootPath = path.join(path.dirname(thisPath), "..");
+const distPath = path.join(rootPath, "dist");
 const esmSourcePath = path.join(rootPath, "index.ts");
-const esmDistPath = path.join(rootPath, "dist", "index.js");
-const commonJsDistPath = path.join(rootPath, "dist", "index.cjs");
+const esmDistDir = path.join(distPath, "esm");
+const esmDistPath = path.join(esmDistDir, "index.js");
+const commonJsDistDir = path.join(distPath, "cjs");
+const commonJsDistPath = path.join(commonJsDistDir, "index.js");
+const typesDistDir = path.join(distPath, "types");
 
 const compileEsm = () =>
-  writeRollup(
-    {
-      input: esmSourcePath,
-      plugins: [rollupTypescript({ tsconfig: "./tsconfig-esm.json" })],
-    },
-    { file: esmDistPath }
+  withEsmFile(esmSourcePath, (esmTempPath) =>
+    writeRollup(
+      {
+        input: esmTempPath,
+        plugins: [rollupTypescript({ tsconfig: "./tsconfig-esm.json" })],
+      },
+      { file: esmDistPath }
+    )
   );
 
 const compileCommonjs = () =>
-  withCommonJsFile(esmSourcePath, (commonJsSourcePath) =>
+  withCommonJsFile(esmSourcePath, (commonJsTempPath) =>
     writeRollup(
       {
-        input: commonJsSourcePath,
+        input: commonJsTempPath,
         plugins: [rollupTypescript({ tsconfig: "./tsconfig-commonjs.json" })],
       },
       {
-        exports: "default",
+        exports: "named",
         file: commonJsDistPath,
         format: "cjs",
       }
-    )
+    ).then(async () => await finalizeCommonJs(commonJsDistDir, esmDistDir))
+  );
+
+const finalizeTypes = async () => {
+  await fs.mkdir(typesDistDir);
+  await fs.mkdir(path.join(typesDistDir, "middlewares"));
+  await renameFile(
+    path.join(esmDistDir, "tmp-esm-index.d.ts"),
+    path.join(distPath, "types", "index.d.ts")
   );
+  await moveDir(path.join(esmDistDir, "middlewares"), path.join(typesDistDir, "middlewares"))
```

Comment:
Why can't we just `fs.rename` this? Why do we need this fancy helper?

Reply:
`fs.rename` doesn't handle moving files to directories that don't exist, nor have any sort of recursive renaming AFAIK. So this function was created to recursively traverse a directory tree and create required directories before moving a file.

However, in commit https://github.com/helmetjs/helmet/pull/345/commits/c8e887c97ff4155dd41853066fea7e34cdd5804a I've removed this function in an attempt to simplify things. 

Instead of outputting the `*.d.ts` type files into `dist/esm/*` and then using the `moveDir` helper to get them over to `dist/types/*`. Now when outputting ESM I output the types straight to `dist/types/*` and then only need to worry about moving `dist/index.js` to `dist/esm/index.js`.

We get to the same place as before, but with less code for us to maintain and a little more readable. Happy for feedback!